File: DP3_Problem1.py
"""
Time complexity: O(nlogn) ??
Space complexity: O(n)
Challenges: Failed many times initiallly while using a counter
Executed in Leetcode: Yes
Comments: Given below
"""

class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        # Result for nums=[3,4,2] is Counter({3: 1, 4: 1, 2: 1})
        """
        Alternate route:
        counter = collections.Counter(nums)   
        select, unselect = 0,0
        """
        # Boundary conditions
        if not nums:
            return 0
        max_no = max(nums)
        
        if max_no <= 2: 
            return (sorted(nums)[-1])
        
        # Create 2 DP array
        x = [0]*max_no
        y = [0]*max_no
        for i in range(len(nums)): x[nums[i]-1] += 1   
        y[0] = x[0]; y[1] = max(x[0],2*x[1])

        for i in range(2,max_no):
            y[i] = max(y[i-1],(i+1)*x[i] + y[i-2])
        
        # The last value is the maximum amount accrued
        return y[-1]

File: DP_3_Problem_1.java

//Time Complexity : O(n)
//Space Complexity: O(1)
//Run successfully on leetcode
//Problem : preprocessing the array

import java.util.Arrays;

public class DP_3_Problem_1 {
    public static int deleteAndEarn(int[] nums) {
        if(nums == null || nums.length == 0) return 0;

        int max = 0;
        for(int i = 0 ; i < nums.length ; i++)
        {
            if(nums[i]>max)
            {
                max = nums[i];
            }
        }
        int[] modified = new int[max+1];
        Arrays.fill(modified,0);
        for(int i = 0 ; i < nums.length ; i ++)
        {
            modified[nums[i]] += nums[i];
        }
        int choose =0;
        int notChoose = 0;
        
        for(int i =1 ; i < modified.length ; i++)
        {
            int temp = Math.max(choose,notChoose);
            choose = notChoose + modified[i];
            notChoose = temp;

        }
        return Math.min(choose,notChoose);


    }

    public static void main(String[] args) {
        int[] nums = {3, 4, 2};
        System.out.println(deleteAndEarn(nums));
    }
}
